Medstuff

comma opened the pull request:
Intent of this stuff is mostly groundwork for removal of stasis bags by providing tools and mechanics to handle two (in my opinion) legit uses of stasis bags - non-restartable hearts tanking oxygenation which can't be helped except full fixing, and medbay being overworked and needing to put some folks on pause.

CPR/autopump stuff adresses first, now EMTs have tools to prolong suffering of their victims in p much all situations.
Sleeper is to address second, between 2 cryotubs and 2 sleepers, Medbay has plenty of pause space.

Nitty gritty:

CPR changes. It now will help with blood circulation even if it fails to restart the heart, so there's a reason to do it (especially on higher levels). Fake circulation lasts 20 seconds and is better with Anatom/Medicine skill (highest).

Can now put IV bags etc on rollerbeds and use them as IV stands for buckled people. Click with beaker to add it. Click with empty hand on empty (no buckled guy) bed to remove it. Drag bed on people to hook/unhook.

Damaged livers no longer generate ammonia. It was creating self-perpetuating damage which wasn't intended, it was for diagnostic memes.

Adds auto-CPR device. Goes into suit slot. It won't do mouth-to-mouth or resusticate them, but it will help with circulation. Needs Trained Anatomy skill to use, or risk rib fracture. Basically continuously applies circulation part of CPR. Mapped bunch of them in medbay

Sleepers now have 10x stasis option. Their power usage now also scales with stasis factor greatly, 500 per factor, so ~5000 W at 10x
Autopump in action on roller with IV stand used

![](https://i.imgur.com/rH7kapN.gif)

comma committed:
CPR changes

Someone being dead won't stop you from trying CPR on them now! It was kinda eh since magical way of letting you know they're braindead.

Now failed breath serve as conditions for CPR too (though you'll skip pumping part if their heart is'nt stopped)

Time to perform CPR is affected by skill (Anatomy), 30ish at unskilled, bit under 20 for Trained

CPR now has effect on blood circulation aside from attempt to restart heart completely. It will sort of 'simulate' heart still working, worse than it would if it was actually started. How well it's done is affected by skill, but generally it's worth doing. Effect lasts for 20 seconds. Remember that there's still oxygenation and blood volume to worry about, so it's not all-fix thing.

Adds ability to use IV on rollerbed

Should be no changes for handling of roller in normal mode, without bothering with beakers.

Otherwise it's:
-Click reagent container on bed to add it
-Drag bed on guy to hook him up. Same to unhook
-Clicking on bed will unbuckle dude and unhook him
-Clicking on bed without buckled dude will remove beaker

Gives sleepers 10x stasis option

Also makes them use more power depending on stasis option, 5 KW per factor of stasis. Was told it's not crazy power drain, I have no idea myself!

Removes ammonia buildup effect from liver damage

It was supposed to be mostly cosmetic for diagnosis memes, but it instead results in self-perpetuating damage, no likey.

Adds auto-CPR device

It goes onto suit slot of the victim, and will regulary keep pumping a-la CPR.
It's better at it than an unskilled guy, and automated.

It won't resusticate on its own, only help with circulation.

If used by someone of low skill (below Trained), will most likely break the ribs from improper placement.

Maps auto-cpr thingies onto map

Changelog

Roland410 commented:
Auto-CPR should have a research design so they aren't as finite.

comma commented:
There's 4 of them, I don't forsee them being very popular, and lmao research, at most medical crate maybe, if people loosing every single one becomes a real problem.

Roland410 commented:
Medical crate is fine as well, also if you're going to remove stasis bags, people will HAVE to use them as alternatives.

ghost commented:
@comma Do these auto-CPR devices fit in a backpack/rucksack? Or are they too large for bags?

afterthought2 commented:
Please send general feedback, inquiries, suggestions, etc. to feedback on discord, as always.

afterthought2 commented on code/game/objects/items/devices/auto_cpr.dm:
```diff
@@ -0,0 +1,58 @@
+	if(force || !istype(H) || slot != slot_wear_suit)
+		return .
+	if(H.species.name in list(SPECIES_HUMAN, SPECIES_SKRELL, SPECIES_UNATHI, SPECIES_ALIEN)) //non-humanoids btfo
+		return .
```
Kill the `.`.

afterthought2 commented on code/game/objects/items/devices/auto_cpr.dm:
```diff
@@ -0,0 +1,58 @@
+
+/obj/item/auto_cpr/Destroy()
+	STOP_PROCESSING(SSobj,src)
+	. = ..()
```
This `STOP_PROCESSING` is handled at `obj` level already.

afterthought2 replied:
Well, just kill the override then.
EDIT: weird comment placement; should be in the newer diff where the line above this one is omitted.

afterthought2 commented on code/game/objects/items/devices/auto_cpr.dm:
```diff
@@ -0,0 +1,58 @@
+
+/obj/item/auto_cpr/Process()
+	if(!ishuman(loc))
+		STOP_PROCESSING(SSobj,src)
```
Better to return `PROCESS_KILL`, so that it's subsystem-agnostic.

afterthought2 commented on code/game/objects/items/devices/auto_cpr.dm:
```diff
@@ -0,0 +1,58 @@
+/obj/item/auto_cpr/equipped(mob/user, slot)
+	..()
+	if(slot != slot_wear_suit)
+		user_skill = user.get_skill_value(SKILL_ANATOMY)
```
I don't think `user` here is the one who is equipping: I think it's just the mob.

comma replied:
Yeah that's who I need. Basically whoever grabbed kit last (`if(slot != slot_wear_suit)`) before putting it on chest counts as one whose skill is checked.

afterthought2 replied:
Maybe I phrased poorly: I think user is the one who is being equipped, not the one doing the equipping. So the victim in this case, not the torturer.

afterthought2 commented on code/modules/mob/living/carbon/human/human_attackhand.dm:
```diff
@@ -76,51 +76,59 @@
+					var/datum/gas_mixture/breath = H.get_breath_from_environment()
+					var/fail = L.handle_breath(breath, 1)
+					if(!fail)
+						to_chat(src, "<span class='notice'>You feel a breath of fresh air enter your lungs. It feels good.</span>")
```
Am confused by the diff. Wasn't all this after the stat check, and that check is now removed? Is that really intended?

comma replied:
Yeah it's intended. Stat check was preventing you from TRYING at all. Now it only prevents resusticatino from happening, os you can still CPR corpses all you want

afterthought2 commented on code/modules/organs/blood.dm:
```diff
@@ -273,21 +272,32 @@ proc/blood_splatter(var/target,var/datum/reagent/blood/source,var/large,var/spra
 /mob/living/carbon/human/proc/get_blood_circulation()
 	var/obj/item/organ/internal/heart/heart = internal_organs_by_name[BP_HEART]
+	if((status_flags & FAKEDEATH) || BP_IS_ROBOTIC(heart))
+		return 100
```
Wait, why do robotic hearts bypass blood total check now?

comma replied:
oh whoops, meant to bypass pulse checking, but they still need to be affected by heart damage, will fix.

afterthought2 reviewed (changes requested):
I believe ironing boards are roller subtypes, so you'll need to kill the IV function on them.

Crushtoe commented on code/game/objects/structures/stool_bed_chair_nest/bed.dm:
```diff
@@ -193,23 +193,114 @@
+		queue_icon_update()
+		
+/obj/structure/bed/roller/proc/remove_beaker(mob/user)
+	to_chat(user, "You dettach \the [beaker] to \the [src].")
```
detach, not dettach

Crushtoe commented on code/modules/mob/living/carbon/human/human_attackhand.dm:
```diff
@@ -76,51 +76,59 @@
+					to_chat(H, "<span class='warning'>You need to remove \the [src]'s mouth covering for mouth-to-mouth resustication!</span>")
+					return 0
+				if (!H.internal_organs_by_name[H.species.breathing_organ])
+					to_chat(H, "<span class='danger'>You need lungs for mouth-to-mouth resustication!</span>")
```
I know this is just copy paste but the proper spelling for this plus above is resuscitation

afterthought2 reviewed (commented):
Looks OK besides the ironing board thing.

afterthought2 reviewed (dismissed):
The roller/ironing board thing is pretty terrible, but i guess I'm letting it slide.

comma commented:
Adjusted sprites to include a board on other side
Made it use medical and anatomy skill (basic) instead of trained anatomy becaus apparently anatomy is more surgeon oriented stuff, I misunderstood.
Added a codex entry for autocpr thingie.

babydoll commented:
Only to be merged alongside this stuff: https://github.com/Baystation12/Baystation12/milestone/19

afterthought2 commented:
Change of plans; merging this first as discussed.
